crates/subspace-service/src/dsn.rs

```diff
@@ -85,17 +89,39 @@ where
             dsn_config.bootstrap_nodes,
         )
         .boxed(),
-        request_response_protocols: vec![PieceByHashRequestHandler::create(move |req| {
-            let result = match piece_cache.get_piece(req.piece_index_hash) {
-                Ok(maybe_piece) => maybe_piece,
-                Err(error) => {
-                    error!(piece_index_hash = ?req.piece_index_hash, %error, "Failed to get piece from cache");
-                    None
-                }
-            };
-
-            Some(PieceByHashResponse { piece: result })
-        })],
+        request_response_protocols: vec![
+            PieceByHashRequestHandler::create(move |req| {
+                let result = match piece_cache.get_piece(req.piece_index_hash) {
+                    Ok(maybe_piece) => maybe_piece,
+                    Err(error) => {
+                        error!(piece_index_hash = ?req.piece_index_hash, %error, "Failed to get piece from cache");
+                        None
+                    }
+                };
+
+                Some(PieceByHashResponse { piece: result })
+            }),
+            RootBlockBySegmentIndexesRequestHandler::create(move |req| {
+                let internal_result = req
+                    .segment_indexes
+                    .iter()
+                    .map(|segment_index| root_block_cache.get_root_block(*segment_index))
+                    .collect::<Result<Vec<Option<RootBlock>>, _>>();
+
+                let result = match internal_result {
+                    Ok(root_blocks) => root_blocks,
+                    Err(error) => {
+                        error!(%error, "Failed to get root blocks from cache");
+
+                        Vec::new()
+                    }
+                };
+
+                Some(RootBlockResponse {
+                    root_blocks: result,
```

Comment:
It is a bit odd to return an empty vector here rather than an error to be honest.

Reply:
I had an idea that it will capture reviewers' attention, however, think about the alternative. IMHO it's confusing to return a peer error here. We have a p2p system - we don't need peers` errors. We could return a fact of the error as a boolean but we still don't need it. Probably, if we return an optional result vector it could make more sense. 

WDUT about `root_blocks: Option<Vec<Option<RootBlock>>>` ?

Reply:
What if we return just `None` here in case of error, isn't that why we have `Option<Response>` here?

Reply:
Makes more sense.